modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/volatility/smile/function/SABRHaganVolatilityFunctionProvider.java

```diff
@@ -0,0 +1,913 @@
+    double strike = option.getStrike();
+    double alpha = data.getAlpha();
+    double beta = data.getBeta();
+    double rho = data.getRho();
+    double nu = data.getNu();
+    if (alpha == 0.0) {
+      return 0.0;
+    }
+    double cutoff = forward * CUTOFF_MONEYNESS;
+    double k;
+    if (strike < cutoff) {
+      Logger s_logger = LoggerFactory.getLogger(SABRHaganVolatilityFunctionProvider.class);
+      s_logger.info("Given strike of {} is less than cutoff at {}, therefore the strike is taken as {}", new Object[] {
+        strike, cutoff, cutoff });
+      k = cutoff;
+    } else {
+      k = strike;
+    }
+    double vol, z, zOverChi;
+    double beta1 = 1 - beta;
+    if (DoubleMath.fuzzyEquals(forward, k, ATM_EPS)) {
+      double f1 = Math.pow(forward, beta1);
+      vol = alpha *
+          (1 + timeToExpiry *
+              (beta1 * beta1 * alpha * alpha / 24 / f1 / f1 + rho * alpha * beta * nu / 4 / f1 + nu * nu *
+                  (2 - 3 * rho * rho) / 24)) / f1;
+    } else {
+      if (DoubleMath.fuzzyEquals(beta, 0, BETA_EPS)) {
+        double ln = Math.log(forward / k);
+        z = nu * Math.sqrt(forward * k) * ln / alpha;
+        zOverChi = getZOverChi(rho, z);
+        vol = alpha * ln * zOverChi *
+            (1 + timeToExpiry * (alpha * alpha / forward / k + nu * nu * (2 - 3 * rho * rho)) / 24) / (forward - k);
+      } else if (DoubleMath.fuzzyEquals(beta, 1, BETA_EPS)) {
+        double ln = Math.log(forward / k);
+        z = nu * ln / alpha;
+        zOverChi = getZOverChi(rho, z);
+        vol = alpha * zOverChi * (1 + timeToExpiry * (rho * alpha * nu / 4 + nu * nu * (2 - 3 * rho * rho) / 24));
+      } else {
+        double ln = Math.log(forward / k);
+        double f1 = Math.pow(forward * k, beta1);
+        double f1Sqrt = Math.sqrt(f1);
+        double lnBetaSq = Math.pow(beta1 * ln, 2);
+        z = nu * f1Sqrt * ln / alpha;
+        zOverChi = getZOverChi(rho, z);
+        double first = alpha / (f1Sqrt * (1 + lnBetaSq / 24 + lnBetaSq * lnBetaSq / 1920));
+        double second = zOverChi;
+        double third = 1 + timeToExpiry * (beta1 * beta1 * alpha * alpha / 24 / f1 +
+            rho * nu * beta * alpha / 4 / f1Sqrt + nu * nu * (2 - 3 * rho * rho) / 24);
+        vol = first * second * third;
+      }
+    }
+    //There is nothing to prevent the nu * nu * (2 - 3 * rho * rho) / 24 part taking the third term, and hence the volatility negative
+    return vol;
+    // return Math.max(0.0, vol);
+  }
+
+  /**
+   * Computes the volatility sensitivity to the SABR parameters. 
+   * <p>
+   * This returns an array with alpha, beta, rho and nu sensitivities.
+   * 
+   * @param option  the option.
+   * @param forward  the forward value of the underlying
+   * @param data  the SABR data.
+   * @return the sensitivities
+   */
+  @Override
+  public DoubleArray getVolatilityModelAdjoint(EuropeanVanillaOption option, double forward, SABRFormulaData data) {
```

Comment:
For adjoint versions, we should now use the "ValueDerivatives" as output object.